File: GoProject/21days/presum/303.go
package presum

type NumArray struct {
	presum []int
	nums   []int
}

func Constructor(nums []int) NumArray {
	presum := make([]int, 0)
	acc := 0
	presum = append(presum, acc)
	for i := 0; i < len(nums); i++ {
		acc = acc + nums[i]
		presum = append(presum, acc)
	}
	return NumArray{
		presum: presum,
		nums:   nums,
	}
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.presum[right+1] - this.presum[left]
}

File: PyProject/leetcode/21/presum/303.py
from typing import List
from functools import reduce


class NumArray:
    def __init__(self, nums: List[int]):
        self.nums = nums

    def sumRange(self, left: int, right: int) -> int:
        return sum(self.nums[left:right + 1])


if __name__ == '__main__':
    a = [1, 2, 3, 4, 5]
    b = reduce(lambda x, y: x + y, a)
    print(b)
